### PR TITLE
Add put_multipart_opts (#5435)

### LINKED ISSUE #5435: Add ObjectStore::put_multipart_opts
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Following #4984 added `ObjectStore::put_opts` taking a `PutOptions` containing a `PutMode`. #4999 extended this to support specifying a `TagSet`. #5334 proposes extending this to support broader categories of metadata.

Whilst `PutMode` does not apply to `put_multipart`, as no stores AFAIK support conditional multipart uploads, but `TagSet` and `Attributes` would be applicable to multipart uploads, as pointed out by @Xuanwo on https://github.com/apache/arrow-rs/pull/5431#discussion_r1502309316

**Describe the solution you'd like**

I would propose introducing a new `PutMultipartOptions` and a corresponding `ObjectStore::put_multipart_opts`, with a default implementation of `ObjectStore::put_multipart` calling through with `PutMultipartOptions::default()`.

**Describe alternatives you've considered**

We could not do this, however, this would create an unfortunate bifurcation in functionality between multipart and non-multipart uploads.

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #5435

# Rationale for this change
 

This allows setting tags and attributes for multipart uploads

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: object_store/src/azure/client.rs [hunk 1]
```diff
@@ -306,27 +310,34 @@ impl AzureClient {
...
             .into_iter()
             .map(|part| BlockId::from(part.content_id))
             .collect();
 
         let payload = BlockList { blocks }.to_xml().into();
         let response = self
-            .put_request(path, payload, Attributes::default())
+            .put_request(path, payload)
+            .with_attributes(opts.attributes)
+            .with_tags(opts.tags)
```
Comment (tustvold) [inline object_store/src/azure/client.rs:338]: Unlike AWS (and GCP which copies AWS) the tags and attributes are specified in the completion request that assembles the object

File: object_store/src/aws/client.rs [hunk 2]
```diff
@@ -335,81 +396,19 @@ impl S3Client {
...
-                    builder.header(CONTENT_TYPE, v.as_ref())
-                }
-            };
-        }
-
-        if !has_content_type {
-            if let Some(value) = self.config.client_options.get_content_type(path) {
-                builder = builder.header(CONTENT_TYPE, value);
-            }
-        }
```
Comment (tustvold) [inline object_store/src/aws/client.rs:380]: Previously we were specifying these for put_part uploads which is technically not correct (although appears to be harmless)

File: object_store/src/aws/client.rs [hunk 3]
```diff
@@ -513,41 +512,29 @@ impl S3Client {
...
-            builder,
-            path: from,
-            config: &self.config,
-            payload: None,
-            payload_sha256: None,
-            use_session_creds: false,
-            idempotent: false,
-        }
+        self.request(Method::PUT, to)
+            .idempotent(true)
```
Comment (tustvold) [inline object_store/src/aws/client.rs:518]: CopyObject is idempotent

File: object_store/src/memory.rs [hunk 4]
```diff
@@ -503,10 +509,11 @@ impl MultipartUpload for InMemoryUpload {
...
         let parts = self.parts.iter().flatten();
         parts.for_each(|x| buf.extend_from_slice(x));
-        let etag = self
-            .storage
-            .write()
-            .insert(&self.location, buf.into(), Attributes::new());
+        let etag = self.storage.write().insert(
+            &self.location,
+            buf.into(),
+            std::mem::take(&mut self.attributes),
```
Comment (tustvold) [inline object_store/src/memory.rs:515]: Calling complete multiple times is documented as implementation defined behaviour so this is fine

File: object_store/src/aws/client.rs [hunk 5]
```diff
@@ -289,11 +286,79 @@ impl<'a> Request<'a> {
...
+            }
+        }
+        Self { builder, ..self }
+    }
+
+    pub fn with_payload(self, payload: PutPayload) -> Self {
+        let mut builder = self.builder;
+
+        let mut sha256 = Context::new(&digest::SHA256);
+        payload.iter().for_each(|x| sha256.update(x));
```
[the remark this answers is not in the document]
Comment (tustvold) [reply object_store/src/aws/client.rs:344]: This was a bug :sweat_smile: 

We need to calculate this here for request signing

**Review by crepererum — APPROVED** [review]

**Pull request merged**